Journey Escape Final Submission

jiasuni opened the pull request:
# Journey Escape – JAXAtari Game Implementation

### Game Overview

**Journey Escape** is a vertically scrolling obstacle-avoidance game where the player must survive against moving obstacles while managing score, time, and power-ups.

Game ends when the timer reaches zero and automatically resets

---

### Obstacles & Items

Obstacles spawn in rows and move downward continuously.
 
| Type         | Behavior                                  | Damage/Bonus |
| ------------ | ----------------------------------------- |--------------|
| Stage Barrier        | Solid, blocks movement                    |       0      |
| Loyal Roadie | Grants temporary invincibility            |       0      |
| Love-Crazed Groupies | Score penalty on contact                  |     -300     |
| Shifty-Eyed Promoter | Large score penalty                       |    -2000     |
| Photographer | Blinks between solid and ghost states     |     -300     |
| Mighty Manager    | Grants effectively infinite invincibility |    +9900     |

---

### Implemented Modifications
#### Basic Mods
- Static background
- Speed up player
- Speed up obstacle
- Reduce player size
- Restrict player movement to only up, down, left and right

#### Complex Mods
- Obstacle diagonal movement (Level 2)
- Obstacle steeper diagonal movement (Level 3)
- Obstacle accelerating movement (Level 4)
- Obstacles random direction (Level 5)
- Chaotic obstacle movement (Extra)

### Other Modification Ideas (Not implemented here)
- Tune animation speed
- Hide specific areas of the playing field 
- Black-and-White mode
- Wraparound playfield (when you leave one edge, you reappear on the opposite edge)

Commit:
copied freeway and renamed it

python .\scripts\play.py -g Journey_Escape now plays freeway_copy

Commit:
added horizonal and diagonal player movement

Commit:
adjusted collision

M4GI3R committed:
Merge pull request #13 from M4GI3R/2-extend-player-movement

2 extend player movement

M4GI3R committed:
Merge pull request #14 from M4GI3R/journey_escape

Journey escape

Commit:
simplification

removed everything related to cars, collision, losing lives, cooldown

Commit:
always walking animation + removed 2nd player

Commit:
changed classnames

jiasuni committed:
Merge pull request #15 from M4GI3R/1-obstacle_lane_rotate_

Simplification / code cleanup

Commit:
added obstacles

jiasuni committed:
Merge pull request #20 from M4GI3R/5-add_obstacles

added obstacles

Commit:
changed game layout

Commit:
remove respawn and adjust player speed

Commit:
added collision detection
added score decrease when collision detected
removed respawn when hitting top
removed scoring when hitting top

Commit:
comment cleanup

Commit:
add right border in background and adjust player position

Commit:
added player speed
added no horizonal movement to get behind obstacle allowed

Commit:
adjust obstacles and player max positions

jiasuni committed:
Merge pull request #21 from M4GI3R/16-add_collision_detection

Add collision detection

Commit:
Merge remote-tracking branch 'origin/journey_escape' into 18-change_game_layout

Commit:
Rename player_walk_straight0.npy, player_walk_straight1.npy to player_walk_front_0.npy,  player_walk_front_1.npy

Commit:
adjust player filename in Code

Commit:
Delete old player sprites

Commit:
add player sprites for left and right movement

Commit:
remove player speed change

Commit:
Clean up leftover comment

Commit:
added obstacle speedUp when player hits top border

jiasuni committed:
Merge pull request #22 from M4GI3R/18-change_game_layout

18 change game layout

Commit:
adjusted observation information

M4GI3R committed:
adjusted observation information #24

adjusted observation information

Commit:
removed non game related edits

Commit:
removed edits in README.md

Commit:
Added obstacle and updated player sprites, adjusted related code logic

PaulSeitz commented:
Hey,

I took a look at your implementation of Journey Escape and wanted to give some (minor) feedback.

The technical aspect, i.e. jax compatibility and framework integration, are great! (see the test GitHub action results)

The basic gameplay mechanics as they are right now look good as well, altough obviously not yet finetuned in terms of speed (ALE is slower).
A minor thing, the ALE version of the game for some reason has a slightly differently sized image output (160, 230, 3), i.e. the ALE image is stretched. Therefore your current image is slightly different from the baseline (which is always the ALE version!).

All other things that as far as I can see are currently missing are listed by you in the comment. So overall, good job so far!

Regarding your questions:

1. 
For most of the other games that have enemy wave systems (like seaquest) we try to find a rough pattern to the randomness (for example that some enemy types always spawn after others, 
or that the number of enemies changes in specified patterns) but if that is not possible we use random numbers. So since you will have played the game way more than I have, I think its best if you decide whats the best way to get a similar challenge level for the agent.
(I played around a bit with different seeds and think that there is a lot of randomness in there, at least in terms of number of enemies. Maybe the enemy types are deterministic, but I am not sure)

2. 
Pretty important, in the best case at least the movement is approximated very closely (as you will see once you train the agent and compare the training progress). 
For me it is easiest to start the game with the ALE_RAMStateDeltas script and go into the frame-by-frame mode (f button, then n to step one frame), from there you can note down the patterns of movement, they probably use a system in which they step and then wait for a determined amount of time to 'fake' movement that is slower than 1px per frame. Then you can reimplement this exact pattern, either by using float values (and only applying the movement changes once an entire pixel was moved) or by copying the 'move-and-wait' pattern from the original atari.
If journey escape has dynamic speed scaling the whole thing gets a bit more complicated but the increases are most probably linear. If they are not, you can approximate it, that would be a case in which 100% approximation is not necessary.

3. 
Interesting, I was not aware that ALE has some games incomplete versions, but if the ALE version has only one level then you only have to do one level. The ALE version always acts as the ground truth, if it lacks something that is described in a manual or part of a different emulation you can ignore it as well.

4. 
Mmmh. Good Question. I think if the background loops through something like 12 different frames, then you could try to copy it? But if its 100 or just random altogether you can randomize it as well. As you say it should be an unnecessary visual effect, but its possible that those effects are a challenge for an agent learning to play it (it shouldnt be, but I never tested it..) and could therefore be interesting for research? I think dont waste too much time on it, but if its easily copied it would be nice to have.
If you decide to use some form of randomization, you can keep the static image in as well (optional, set by a constant?), since I imagine a randomized background will be a performance nightmare for the jax based renderer.

M4GI3R committed:
Sprites updated: obstacle and player 

Added obstacle and updated player sprites, adjusted related code logic

Commit:
- added big obstacles (scaling)
- added blinking of lightbulb obstacle
- added scripted spawn of blue robots as 3rd obstacle (like in ALE)
- reworked side- , top- and bottom bar
  - obstacles spawn and despawn behind top and bottom bar
  - they are independent of the background sprite and drawn on top
  - changed the blue color of top and bottom bar to match ALE

  ToDo:
  - define enemy sequences to avoid bad RNG
  - vertical spacing variations
  - fix hitboxes (probably new branch)
  - check if lightbulb offtime = ontime in ALE

Commit:
- added probabilities like in ALE episode

M4GI3R committed:
Enhance Obstacles [1/2]

Enhance Obstacles [1/2]

Commit:
rename: all "chicken" to "player" and "car" to "obstacles"

jiasuni committed:
Merge pull request #34 from M4GI3R/26-remove_all_chicken_references

rename: all "chicken" to "player" and "car" to "obstacles"

Commit:
Timer as 59sec countdown and sprites added

Commit:
reset game if timer ends

jiasuni committed:
Merge pull request #35 from M4GI3R/11-add_timer

11 add timer

Commit:
Convert score to dollars and recolored digit sprites

Commit:
fix countdown rendering

Commit:
update existing obstacle sprites and add big obstacle sprites

Commit:
Added missing obstacles (big robot, fire face) and renamed "item" to "robot"

Commit:
Remove unused obstacle and old sprites

M4GI3R committed:
Merge pull request #36 from M4GI3R/10-points_to_dollars

10 points to dollars

Commit:
in progress: implement robot logic

Commit:
Merge branch 'journey_escape' into 8-enhance_obstacles-22
merge journey_escape

Commit:
update player sprites; TODO: adjust hitbox for collision

jiasuni committed:
Merge pull request #37 from M4GI3R/32-sprites_overhaul

update player sprites; TODO: adjust hitbox for collision

Commit:
- added new player sprite from journey_escape

Commit:
- added new player dimensions

Commit:
- fixing player hitbox

Commit:
Merge branch 'journey_escape' into 8-enhance_obstacles-22

# Conflicts:
#	src/jaxatari/games/jax_journey_escape.py

Commit:
- collision refactor
- type based scoring
- invincible effect
- speed calibration

Commit:
add more sprites for background

jiasuni committed:
Merge pull request #38 from M4GI3R/8-enhance_obstacles-22

Obstacles Enhancement

Commit:
filename correction

Commit:
filename correction

Commit:
merge current journey_escape

Commit:
added variable background-animation speed (option to have static background)

M4GI3R committed:
Merge pull request #39 from M4GI3R/31-enhance_background_v2

added variable background-animation speed

(option to set to 0 for static background)

Commit:
update image height 210 -> 230

Commit:
adjusted background and sprite positions

jiasuni committed:
Merge pull request #44 from M4GI3R/42-image-size

42 image size

Commit:
Rename game elements to match game doc

jiasuni committed:
Merge pull request #45 from M4GI3R/40-updates-based-on-game-doc

Rename game elements to match game doc

jiasuni committed:
Merge branch 'k4ntz:master' into journey_escape_update

Commit:
refactor constants class

Commit:
updated states, observations, and info as struct.dataclasses

jiasuni committed:
Merge pull request #47 from M4GI3R/journey_escape_update

Journey escape update

Commit:
add mod setup and intial integration

jiasuni committed:
Merge pull request #48 from M4GI3R/46-add-mods-setup

add mod setup and intial integration

Commit:
Add 2 Basic Mods: BackgroundStatic, RestrictPlayerSize

Commit:
Add basic mod ReducePlayerSize

Commit:
fixed failing pytests

Commit:
Merge branch 'journey_escape' into 49-basic-mods

Commit:
implemented Level 2 mod

Commit:
slower horizontal speed + black area bottom fix

Commit:
added L3 L4 L5

jiasuni committed:
Merge pull request #51 from M4GI3R/49-basic-mods

49 basic mods

Commit:
fixed black area bottom

Commit:
added chaos mode, fixed L5

Commit:
Merge remote-tracking branch 'origin/journey_escape' into complex_mods

# Conflicts:
#	src/jaxatari/games/jax_journey_escape.py

Commit:
- Added double speed (2 px/frame) for Down, Left, and Right directions when FIRE (Spacebar) is pressed
- Refined walking animation so Up-diagonal movements correctly prioritize the front-facing sprite
- Updated movement animation to fall back to the front-facing sprite when blocked by walls like in the original ALE

jiasuni committed:
Merge pull request #52 from M4GI3R/complex_mods

Added complex mods

jiasuni committed:
Merge pull request #54 from M4GI3R/53-add-spacebar-feature

feat: implement FIRE button sprint and correct sprite dominances

Commit:
change prob of mighty manager

Commit:
updated action space

M4GI3R committed:
change-action-space

updated action space

Commit:
added observations
added end Episode on Score 0
removed all debug prints

M4GI3R committed:
ready-for-training

Commit:
Remove changes from core.py for final submisssion

Commit:
Minor comment updates
